Fixed "No such file or directory" error when  configuring "vm_set_proxy"

pretty66 opened the pull request:
This PR is used to fix the environment variable "VM_SET_PROXY" that was set when building the development environment, and an error occurred: "No such file or directory" error problem

Fixes: #19131

tavif committed:
development: Installation script error

When I set up the development environment, I need to set the agent for some reasons. When I configure the environment variable "vm_set_proxy" according to the document, an error occurs:
$ cilium git:(master) ✗ K8S=1 NO_BUILD=1 VM_MEMORY=8192 VM_CPUS=4 VM_SET_PROXY=http://my.proxy.com:1080 ./contrib/vagrant/start.sh
...
k8s1: /home/vagrant/go/src/github.com/cilium/cilium/contrib/vagrant/scripts/helpers.bash: line 106: HTTPS_PROXY=http://my.proxy.com:1080: No such file or directory
...

This PR is used to fix the environment variable "VM_SET_PROXY" that was set when building the development environment, and an error occurred: "No such file or directory" error problem

Fixes: #19131

Signed-off-by: xujp <1240402516@qq.com>

tavif committed:
development: Installation script error

When I set up the development environment, I need to set the agent for some reasons. When I configure the environment variable "vm_set_proxy" according to the document, an error occurs:
$ cilium git:(master) ✗ K8S=1 NO_BUILD=1 VM_MEMORY=8192 VM_CPUS=4 VM_SET_PROXY=http://my.proxy.com:1080 ./contrib/vagrant/start.sh
...
k8s1: /home/vagrant/go/src/github.com/cilium/cilium/contrib/vagrant/scripts/helpers.bash: line 106: HTTPS_PROXY=http://my.proxy.com:1080: No such file or directory
...

This PR is used to fix the environment variable "VM_SET_PROXY" that was set when building the development environment, and an error occurred: "No such file or directory" error problem

Fixes: #19131

Signed-off-by: xujp <1240402516@qq.com>

aditighag reviewed (changes requested):
Thanks for the PR. Could you add more details to the commit description - where the issue was coming from? Are you looking to set up proxy in the vagrant VM?

joestringer commented on contrib/vagrant/scripts/helpers.bash:
```diff
@@ -103,7 +103,8 @@ function download_to {
 
         rm -f "/tmp/${component}"
-        ${WGET} -O "/tmp/${component}" -nv "${url}"
+        # fix: No such file or directory
+        eval ${WGET} -O "/tmp/${component}" -nv "${url}"
         # Hide 'failed to preserve ownership' error
         mv "/tmp/${component}" "${cache_dir}/${component}" 2>/dev/null
```
This code looks like it is intended to download a component and place it in a certain directory. I don't understand how eval makes a difference. Can you describe the intent here?

pretty66 replied:
The "eval" command will first scan the command line for all substitutions, and then execute the command. fixed "No such file or directory" error when configuring "vm_set_proxy”

You can simply reproduce this error problem by executing the following command to simulate the error
```shell
cat >> test.sh << 'EOF'
export WGET="HTTPS_PROXY=http://you.proxy.domain wget"
${WGET} https://www.google.com
EOF
chmod +x test.sh
./test.sh
```

joestringer replied:
Is it possible to solve this without `eval`?

I was wondering perhaps whether it's sufficient to add a line something like this to the top of the file:

```
HTTPS_PROXY={HTTPS_PROXY:-""}
```

Or, similar to the other scripts, change the WGET commandline based on the VM_PROXY variable?

pretty66 replied:
Currently testing adding eval here can solve this problem with minimal changes to the code

joestringer replied:
^^ Mostly I'm just looking at the original PR https://github.com/cilium/cilium/pull/7289 and wondering why the `eval` is necessary. This seems like the code is intended for the caller to pass in a `WGET` variable so that we don't have to `eval` here. Maybe @jomenxiao could chime in on how they made this work before and what the best solution is?

pretty66 committed:
development: Installation script bug

When setting up a development environment, for some reason I need to set up a proxy. When I configure the environment variable "vm_set_proxy" according to the documentation, wget is not using the proxy correctly

This PR fixes the bug that the proxy setting of the wget command does not take effect when creating a development environment

Signed-off-by: xujp <1240402516@qq.com>

pretty66 committed:
Merge branch 'fix-development-env' of github.com:pretty66/cilium into fix-development-env

Signed-off-by: xujp <1240402516@qq.com>

joestringer commented on contrib/vagrant/start.sh:
```diff
@@ -284,7 +284,7 @@ if [ -z "${RELOAD}" ]; then
 
 if [ -n "${VM_PROXY}" ]; then
-    export WGET="HTTPS_PROXY=${VM_PROXY} wget"
+    export WGET="https_proxy=${VM_PROXY} wget"
```
I'm surprised that this change was necessary, but cross-referencing with the `wget` manpages this does seem to be correct. I don't know why the original author of this code didn't notice this. :+1:
